### ISSUE TITLE
Selectors: Generalized vs. Specific use case

### ISSUE THREAD

**Issue by pscohn** [rainforestapp/frontend-conventions#18]
It seems there's a couple potential ways to define selectors that I've seen in the codebase, and I've had to think about when getting new stuff from state. I thought we could talk about the pros/cons of each way and see if it makes sense to standardize it across the app one way or the other.

### 1. create generalized selectors that accept an arbitrary path to find some property:

```
export const selectClient = ({ client }: State, path: string[]) => client.getIn(path);
```

In use:

```
failPopup: selectClient(state, ['settings', 'fail_test_on_popup']),
```

**Pros:**
- requires fewer, more reusable functions
- requires less boilerplate

**Cons:**
- Forces the caller to know the exact path and names required
- potential duplication in arguments across files (might require multiple changes if state structure changes)

### 2. Create specific selectors for each use that take no extra parameters

With this approach you'd need to create a new selector for each separate property you're selecting in mapStateToProps. for example:

```
export const selectFailPopupSetting = ({ client }: State) =>
  client.getIn(['settings', 'fail_test_on_popup'])';
// can also be refactored to call an internal, general select method like the one above)
```

In use:

```
failPopup: selectFailPopupSetting(state)
```

**Pros:**
- fewer arguments required - generally just `state`, and sometimes `id`
- components don't have to care about how to get something or the path of an object, or what the default values should be
- only have to change code in selectors to change any values derived from state, including default/empty values

**Cons:**
- requires more functions - one for each use case in `mapStateToProps` - and more code to test, more imports

Maybe you can tell from my number of pros and cons, but I feel the second option is more aligned with some of the goals when we decided to start using selectors/reselect. Part of it was to make sure components don't have to know anything about the internal representation of the state, and with the first option combined with some logic that still exists in mapStateToProps, we're not doing that as much as we could. (In the future we should probably also talk about `.toJS()` :trollface:) 

I'd love to hear your take, more pros/cons you can think of @rainforestapp/frontend

**Comment by AJFunk** [comment]
I like #2 as well. Selectors become useful when they are specific. Using a generalized selectors seems to take away from the value of using reselect in the first place. This does mean having a lot more selectors, but I don't think that's a problem.

Addressing `.toJS()`, it would be really great if every selector returned normal JS objects so we wouldn't have to deal with immutable structures inside of components. I wonder if there's a good way to configure all selectors to just do `.toJS()` for us? If so, it would require a sizable refactor... but maybe we could start doing it manually inside of each selector?

**Comment by jkaipr** [comment]
I've been doing what you described as `#2` iirc. Isolation of state access into one place seems like main benefit to me.

**Comment by pscohn** [comment]
@AJFunk maybe we should have a separate discussion about toJS, but from what I've read in the redux docs[1] it's better to avoid using `toJS` at all as it's rather slow, and if used for rendering can have pretty bad performance. But I'm not sure if there's any pattern to how we use it so far (it seems like we use it sometimes when calling actions with objects)

[1] http://redux.js.org/docs/recipes/UsingImmutableJS.html

**Comment by ziahamza** [comment]
I think we can expose both, a general selector for an entity and
specialised selectors. Some fields are only used once, so no point making a
dedicated selector for them.

If it's repeated more then a couple of times then make a specialised
selector

**Comment by prayogoa** [comment]
I want to address `Forces the caller to know the exact path and names required` as a con.

I think at some point we will have to be able to make assumptions about the shape of the data we are operating with to be able to do useful things with it. To me I draw the line at having to know where in a model is the information I need (eg. knowing that a client's fail popup preference is under the path `settings.fail_test_on_popup` is as much knowledge burden as knowing to call the `selectFailPopupSetting` selector) as opposed to having to know where in our state structure is that model located ( the client model is located in `state.client`). Because of this I wouldn't necessarily consider this a major con of the first method. 

`only have to change code in selectors to change any values derived from state, including default/empty values` is however a valid and major con of the second method for me, especially if it includes actual derivation instead of just straight up reading from the state.

What about this:
1) Start with using the general selector
2) The moment the same path is used twice, refactor into a specialized selector
3) If doing any sort of computation / derivation, always use specialized selector

**Comment by prayogoa** [comment]
re `.toJS`, it's an escape hatch to deal with older components that expects mutable objects as props. Imo we should try to keep immutability as much as possible.

**Comment by prayogoa** [comment]
After some further discussions I think the rule I previously proposed before would be hard to follow, since it puts the burden to scour the codebase for uses of the general selector on the second implementor. I'm leaning towards always using specialized selectors now. Each entity has a finite amount of attributes anyway so I'm not expecting an exploding number of selector functions if we do that. It's also much easier to scan the individual selector file for refactor opportunities than to grep the entire codebase.

Open question: Is there a use case for dynamic pathing in selectors? We've talked about using generalized selectors to reduce the number of functions, but generalized selectors also allow us to dynamically access entities, this isn't something the specialized selectors can do.

**Comment by jkaipr** [comment]
There are places where we select and return the whole entity e.g. `selectTest(state, testId)`. How should we handle these situations?

They allow for using the related entity state in any way we please, including dynamic access to it's values. Are we ok with using those? It means that if you decide, you don't have to use the specialized selectors. If we forbid their usage, every attribute would have to be selected separately. Not sure if it's practical and it would probably mean some refactoring.

**Comment by jkaipr** [comment]
Discussed this further with Adrian and the outcome was that we could store responses as Immutable [Records](https://facebook.github.io/immutable-js/docs/#/Record). This would allow for much better type checking and Records also support destructuring the same way as plain JS Objects.
This means that the selector which returns the base model of a reducer by would return a flow typed record (e.g. Test, RunGroup, ...) which has a clearly defined shape, unlike our current Maps `Map<string, any>` where the values can be of any type. No need to worry about what I described in my previous comment anymore. This Record takes care of all access to plain values of the model, we would need specialised selectors only for cases where we derive values or connect multiple related models to get a result.

It involves quite a bit of refactoring, but we could do it an entity at a time.

**Comment by prayogoa** [comment]
We couldn't do this before because the typings for records were bad, but they seem to have rewritten it for v4 and it's supposedly a lot better now. https://facebook.github.io/immutable-js/docs/#/Record

**Comment by pscohn** [comment]
It would probably take some time to move to records if we decide to, and that can be a separate discussion. In the short term, should we scrap a specialized selector convention, or go ahead and use it for new code with some exceptions?

**Comment by prayogoa** [comment]
We have static typing improvement in the roadmap so moving to record is something that I definitely plan to do. In the short term I think it would still be beneficial to use specialized selectors and refactor them once we have the records in place. It'll be more work down the line but I think it's worth it

**Comment by pscohn** [comment]
For the static typing conversation, another option is to do use something like immer and get rid of immutable, then we can type all our models much more easily (and reduce bundle size)

https://github.com/mweststrate/immer

**Comment by pscohn** [comment]
In either case, I think the ideal is when

**Comment by pscohn** [comment]
weps

**Comment by prayogoa** [comment]
There's some experimentation with immer going on right now with beak specs, lets open another discussion once we have more info on that to discuss static typing
